net/core/filter.c

```diff
@@ -9142,6 +9162,16 @@ u32 bpf_sock_convert_ctx_access(enum bpf_access_type type,
 		*target_size = 2;
 #endif
 		break;
+
+	case offsetof(struct bpf_sock, cookie):
+		*insn++ = BPF_ATOMIC_LOAD_OP(
```

Comment:
why is this implemented with an atomic load and not the same way as e.g. the `dst_port` case which uses `BPF_LDX_MEM`?

Reply:
John mentioned that we need to make it an atomic load because the underlying value `skc_cookie` is atomic and generated via 

```c
u64 __sock_gen_cookie(struct sock *sk)
{
        while (1) {
                u64 res = atomic64_read(&sk->sk_cookie);
        
                if (res)
                        return res;
                res = gen_cookie_next(&sock_cookie);
                atomic64_cmpxchg(&sk->sk_cookie, 0, res);
        }
}
```

> we should also do an atomic read. in the inline insn

I don't think that's the case for `dst_port`.

Reply:
Ah right. Hm does the xchg in the op-code have any unintended effects? E.g. the cookie won’t get unset?